### PR TITLE
Add validium support for L1 recovery (DAC)

### PR DESCRIPTION
**Pull request by cffls**
This PR adds the support of recovering from validium. When a sequencer is recovering from L1, it will detect if the data on L1 is a data hash (validium mode), and subsequently query data from a DA server if required. 

Example usage:

Add the following flag to yaml config:

```yaml
zkevm.da-url: http://127.0.0.1:36063
```

### CODE REVIEW

**Review by hexoscott — COMMENTED** [review]
Looks good!  

One thing I'm unsure of at the moment though is that we're now looking for a new type of verification which makes sense in an RPC context.  But the L1 block recovery is designed to only run as a sequencer which doesn't run the verifications, and I can't see a new type of contract being used in the L1 block sync process, I'd imagine it would have a different topic to the usual rollup?

File: zk/l1_data/l1_decoder.go [hunk 1]
```diff
@@ -65,9 +88,39 @@ func DecodeL1BatchData(txData []byte) ([][]byte, common.Address, error) {
...
-	if err != nil {
-		return nil, coinbase, err
+
+	if !isValidium {
+		err = json.Unmarshal(bytedata, &sequences)
+		if err != nil {
+			return nil, coinbase, err
+		}
+	} else {
+		var validiumSequences []ValidiumBatchData
```
Comment (mandrigin) [inline zk/l1_data/l1_decoder.go]: I would extract it to a method, already many lines of logic. Ideally even have something like that

```
if isValidium {
    GetSequencesForValidum(…)
} else {
    GetSequencesForRollup(…)
}
```

File: zk/stages/stage_l1syncer.go [hunk 2]
```diff
@@ -230,6 +230,10 @@ func parseLogType(l1RollupId uint64, log *ethTypes.Log) (l1BatchInfo types.L1Bat
 		batchLogType = logVerify
 		batchNum = new(big.Int).SetBytes(log.Topics[1].Bytes()).Uint64()
 		stateRoot = common.BytesToHash(log.Data[:32])
+	case contracts.VerificationValidiumTopicEtrog:
+		batchLogType = logVerify
```
Comment (mandrigin) [inline zk/stages/stage_l1syncer.go]: Shouldn’t we take rollupID into account here as well?

File: zk/txpool/txpooluitl/all_components.go [hunk 3]
```diff
@@ -22,8 +22,8 @@ import (
 	"time"
 
 	"github.com/c2h5oh/datasize"
-	"github.com/holiman/uint256"
 	"github.com/gateway-fm/cdk-erigon-lib/txpool/txpoolcfg"
+	"github.com/holiman/uint256"
```
Comment (mandrigin) [inline zk/txpool/txpooluitl/all_components.go]: Unnecessary change

File: zk/stages/stage_sequence_execute_transactions.go [hunk 4]
```diff
@@ -19,10 +21,9 @@ import (
 	"github.com/ledgerwatch/erigon/core/vm"
 	"github.com/ledgerwatch/erigon/core/vm/evmtypes"
 	"github.com/ledgerwatch/erigon/rlp"
+	"github.com/ledgerwatch/erigon/zk/constants"
 	"github.com/ledgerwatch/erigon/zk/hermez_db"
 	zktx "github.com/ledgerwatch/erigon/zk/tx"
-	"errors"
```
Comment (mandrigin) [inline zk/stages/stage_sequence_execute_transactions.go]: Unnecessary change

**Review by mandrigin — CHANGES_REQUESTED** [review]
A few things to both clarify and fix

File: zk/da/client.go [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+		return nil, err
+	}
+
+	httpReq.Header.Add("Content-type", "application/json")
+
+	return httpReq, nil
+}
+
+func GetOffChainData(ctx context.Context, url string, hash common.Hash) ([]byte, error) {
+	response, err := JSONRPCCallWithContext(ctx, url, "sync_getOffChainData", hash)
```
Comment (zjg555543) [inline zk/da/client.go]: LGTM for all the code.Thank you for supporting the Validium feature.
I'm not sure that if it can be replace this with the existing [JSONRPCCall](https://github.com/0xPolygonHermez/cdk-erigon/blob/v1.1.2/zkevm/jsonrpc/client/client.go#L28) code to make it more concise, and it can loop here for attempts too.

**Comment by cffls** [comment]
> Looks good!
> 
> One thing I'm unsure of at the moment though is that we're now looking for a new type of verification which makes sense in an RPC context. But the L1 block recovery is designed to only run as a sequencer which doesn't run the verifications, and I can't see a new type of contract being used in the L1 block sync process, I'd imagine it would have a different topic to the usual rollup?

Great question @hexoscott ! A new verification we need for validium is to check if the data received from DA matches the hash posted to L1, and it is checked [here](https://github.com/0xPolygonHermez/cdk-erigon/pull/560/files#diff-7377b0f3e9b13c0e3c936eb40406bb5fdd43ccc67623cb2dce9d59f0677216cdR113-R115).
For a RPC, it will receive full data from a data streamer as it is today. Hence no change is required there. For a sequencer to recover from L1, it needs the additional data provided by DA. Luckily, we don't need a new topic for sequence event at the moment, because the validium contract emits the same sequence event as rollup, see code [here](https://github.com/0xPolygonHermez/zkevm-contracts/blob/1ad7089d04910c319a257ff4f3674ffd6fc6e64e/contracts/v2/consensus/validium/PolygonValidiumEtrog.sol#L249C14-L249C29).

File: zk/l1_data/l1_decoder.go (hunk 1 above)
Comment (cffls) [reply zk/l1_data/l1_decoder.go]: Sounds good, will do.

File: zk/stages/stage_l1syncer.go (hunk 2 above)
Comment (cffls) [reply zk/stages/stage_l1syncer.go]: Good catch! Will add the check.

File: zk/da/client.go (hunk 5 above)
Comment (cffls) [reply zk/da/client.go]: Thanks for the pointer! I will see if it could be reused here.

File: zk/da/client.go [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+		}
+
+		if response.Error != nil {
+			return nil, fmt.Errorf("%v %v", response.Error.Code, response.Error.Message)
+		}
+
+		return hexutil.Decode(strings.Trim(string(response.Result), "\""))
+	}
+
+	return nil, fmt.Errorf("max attempts of data fetching reached, attemps: %v, DA url: %s", maxAttempts, url)
```
Comment (rachit77) [inline zk/da/client.go]:
```suggestion
	return nil, fmt.Errorf("max attempts of data fetching reached, attempts: %v, DA url: %s", maxAttempts, url)
```

**Comment by hexoscott** [comment]
> > Looks good!
> > One thing I'm unsure of at the moment though is that we're now looking for a new type of verification which makes sense in an RPC context. But the L1 block recovery is designed to only run as a sequencer which doesn't run the verifications, and I can't see a new type of contract being used in the L1 block sync process, I'd imagine it would have a different topic to the usual rollup?
> 
> Great question @hexoscott ! A new verification we need for validium is to check if the data received from DA matches the hash posted to L1, and it is checked [here](https://github.com/0xPolygonHermez/cdk-erigon/pull/560/files#diff-7377b0f3e9b13c0e3c936eb40406bb5fdd43ccc67623cb2dce9d59f0677216cdR113-R115). For a RPC, it will receive full data from a data streamer as it is today. Hence no change is required there. For a sequencer to recover from L1, it needs the additional data provided by DA. Luckily, we don't need a new topic for sequence event at the moment, because the validium contract emits the same sequence event as rollup, see code [here](https://github.com/0xPolygonHermez/zkevm-contracts/blob/1ad7089d04910c319a257ff4f3674ffd6fc6e64e/contracts/v2/consensus/validium/PolygonValidiumEtrog.sol#L249C14-L249C29).

Great!  Thank you for the clarification there

**Review by mandrigin — DISMISSED** [review]

File: zk/da/client.go (hunk 5 above)
Comment (mandrigin) [reply zk/da/client.go]: See it is used now. Resolving the conversation.

**Review by rachit77 — DISMISSED** [review]

**Review by mandrigin — APPROVED** [review]
